src/Environment.php

```
@@ -395,15 +435,29 @@ public static function provider(): ?ProviderInterface {
-      // Load known providers from the constant (no filesystem scanning).
-      foreach (self::PROVIDERS as $class) {
-        static::addProvider(new $class());
+      static::$providers = [];
+
+      $instances = array_merge(self::PROVIDERS, $additional);
+
+      foreach ($instances as $instance) {
+        $instance = is_string($instance) ? new $instance() : $instance;
+
+        if (!($instance instanceof ProviderInterface)) {
+          throw new \InvalidArgumentException('The provider must implement ProviderInterface');
```

PHPMD (2.15.0)

Missing class import via use statement (line '454', column '21'). (undefined)

(MissingImport)